### PR TITLE
Add Scenario Libraries View

### LINKED ISSUE #176: Add scenario libraries view
**Is your feature request related to a problem? Please describe.**
We currently only support the baseline scenario in the web model, but ask that the user keep this model reflecting the real-world. This makes it difficult for the user to adequately forecast different scenarios and what-ifs (e.g., 'What if we were to release some of the low-risk members of the incarcerated population? What if we were to make space to move more people out of bunked housing and into cells?')

In order to accomplish this, we need a place in the UI for users to view, open, add, or delete their scenarios.

**Describe the solution you'd like**
Creation of the 'Library' modal dialog for viewing the user's scenarios. See mocks below.

**Additional context**
- This is a child issue of issue #168 
- [Current mocks](https://www.figma.com/file/EGJS9njuKHWdFHUS4lrHA4/%F0%9F%9A%80Recidiviz-CJStatus?node-id=910%3A130)

### PR DESCRIPTION
**Pull request by sprad**
Description of the change

* Adding the Scenario Libraries view to allow users to test out different hypothetical actions and see their resulting impacts

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (adjusts configuration to achieve some end related to functionality, development, performance, or security)

## Related issues

Closes #176 

### Development

These boxes should be checked by the submitter prior to merging:

- [x] Manual testing has been performed locally

### Code review

These boxes should be checked by reviewers prior to merging:

- [x] This pull request has a descriptive title and information useful to a reviewer
- [x] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [x] Potential security implications or infrastructural changes have been considered, if relevant

### CODE REVIEW

**Comment by sprad** [comment]
This PR contains all of the visuals and most of the baseline functionality for the Scenario Library.  At the moment, it's hidden behind a feature flag such that it can be merged after review (with one exception that I'll call out).  The remaining work for this feature is listed in this ticket:  https://github.com/Recidiviz/covid19-dashboard/issues/168

File: src/database/index.tsx [hunk 1]
```diff
@@ -204,7 +204,7 @@ export const getScenarios = async (): Promise<Scenario[]> => {
     // We're sorting in memory instead of in the query above due to a limitation
     // in the way our data is currently modeled. See this issue for more details:
     // https://github.com/Recidiviz/covid19-dashboard/issues/253
-    return sortBy(scenarios, "name");
+    return orderBy(scenarios, ["baseline", "name"], ["desc", "asc"]);
```
Comment (sprad) [inline src/database/index.tsx:207]: We always want the baseline scenario to show up first

File: src/database/type-transforms.tsx [hunk 2]
```diff
@@ -5,7 +5,7 @@ import { PlannedRelease } from "../impact-dashboard/EpidemicModelContext";
 import { Facility, ModelInputs, Scenario } from "../page-multi-facility/types";
 
 const timestampToDate = (timestamp: firebase.firestore.Timestamp): Date => {
-  return timestamp.toDate();
+  return timestamp?.toDate() || new Date();
```
Comment (sprad) [inline src/database/type-transforms.tsx:8]: @jessex - this fixes that seeming race condition you and I were discussing earlier.  I don't think this is an acceptable solution and therefore this should be considered a merge blocker.  I'm working on figuring out a better solution now

File: src/design-system/Modal.tsx [hunk 3]
```diff
@@ -12,10 +12,12 @@ export interface Props {
   open: boolean;
   setOpen: (open: boolean) => void;
   trigger?: string | React.ReactElement<any>;
+  height?: string;
+  width?: string;
```
Comment (sprad) [inline src/design-system/Modal.tsx:16]: With these new properties in place we can size our modals to meet the needs of a given context.  I've done this to support the Scenario Library, but we can also use this to resize our current confirmation modals which are a bit awkward in size using the defaults

File: src/page-multi-facility/ScenarioLibraryModal.tsx [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+    fetchScenarios();
+  }, []);
+
+  const copyScenario = (scenarioId: string) => {
+    setScenarios({
+      data: [],
+      loading: true,
+    });
+
+    duplicateScenario(scenarioId).then(() => {
```
Comment (sprad) [inline src/page-multi-facility/ScenarioLibraryModal.tsx:143]: Given that we wait for the `duplicateScenario` to resolve before re-fetching, it's unclear why we are experiencing the seeming race condition mentioned above 🤷

File: src/page-multi-facility/ScenarioSidebar.tsx [hunk 5]
```diff
@@ -124,24 +147,26 @@ const ScenarioSidebar: React.FC<Props> = (props) => {
...
-          <ToggleRow
-            onToggle={() =>
-              handleScenarioChange({ dataSharing: !scenario?.dataSharing })
-            }
-            toggled={scenario?.dataSharing}
-            label="Share data to improve the model"
-            labelHelp="If enabled, your baseline scenario will be made available to Recidiviz and the research community to improve the model and the state of research on the spread of disease in facilities. Any public research will anonymize state and facility names."
-          />
-          <HorizontalRule />
+          <ToggleContainer hidden={!scenario?.baseline}>
```
Comment (sprad) [inline src/page-multi-facility/ScenarioSidebar.tsx:150]: We only show the toggles and impact button for the baseline scenario.  For now at least

File: src/database/index.tsx [hunk 6]
```diff
@@ -6,7 +6,7 @@ import "firebase/firestore";
 
 import createAuth0Client from "@auth0/auth0-spa-js";
 import { format, startOfDay, startOfToday } from "date-fns";
-import { pick, sortBy } from "lodash";
+import { pick, orderBy } from "lodash";
```
Comment (sprad) [inline src/database/index.tsx:9]: `orderBy` is that new 🔥

File: src/database/type-transforms.tsx (hunk 2 above)
Comment (jessex) [reply src/database/type-transforms.tsx:8]: I agree. Thanks for looking into this!

**Comment by sprad** [comment]
I don't know how to do those fancy gif screen captures, but the experience looks like this:

**Trigger the modal by clicking on the folder icon:**

Screen Shot 2020-05-08 at 12 35 43 PM

**Click Duplicate, and then wait a moment:**

waiting

**Congrats, you have a new Scenario copy!**

Screen Shot 2020-05-08 at 12 38 36 PM

File: src/database/type-transforms.tsx [hunk 7]
```diff
@@ -66,7 +66,7 @@ export const buildScenario = (
 
   let scenario: Scenario = documentData;
   scenario.id = document.id;
-  scenario.createdAt = timestampToDate(documentData.createdAt);
+  scenario.createdAt = timestampToDate(documentData.createdAt );
```
Comment (jessex) [inline src/database/type-transforms.tsx:69]: Whoopsies.

File: src/page-multi-facility/ScenarioLibraryModal.tsx [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+
+const IconRecidviz = styled.img`
+  width: 50px;
+  height: 50px;
+  margin: auto;
+`;
+
+const ScenarioLibraryModal: React.FC<Props> = ({ trigger }) => {
+  const [modalOpen, setModalOpen] = useState(false);
+  const [_, dispatchScenarioUpdate] = useScenario();
```
Comment (jessex) [inline src/page-multi-facility/ScenarioLibraryModal.tsx:117]: @sokunthel had dealt with a similar issue for this by adding an exception to the lint rules, `no-unused-vars` in `.eslintrc.json`. If this is a slightly different rule, it should also be whitelisted same as that one.

File: src/page-multi-facility/ScenarioLibraryModal.tsx (hunk 4 above)
Comment (jessex) [reply src/page-multi-facility/ScenarioLibraryModal.tsx:143]: Is something about that particular persistence call eventually consistent? Are we sure it's a race condition (I'm guessing so if it only happens for you on occasion)?

File: src/database/index.tsx [hunk 9]
```diff
@@ -545,7 +545,7 @@ export const duplicateScenario = async (
       });
     }
 
-    batch.commit();
+    await batch.commit();
```
Comment (sprad) [inline src/database/index.tsx:548]: This fixes the "race condition" which makes a lot of sense.  I wasn't waiting for the commit to finish before trying to load the scenarios and therefore data was missing when trying to hydrate the objects in the transforms.  This is indeed the way other similar methods like `saveFacility` are working:  https://github.com/Recidiviz/covid19-dashboard/blob/master/src/database/index.tsx#L438

Comment (jessex) [reply src/database/index.tsx:548]: Great find!

File: src/page-multi-facility/ScenarioLibraryModal.tsx [hunk 10]
```diff
@@ -114,7 +114,7 @@ const IconRecidviz = styled.img`
 
 const ScenarioLibraryModal: React.FC<Props> = ({ trigger }) => {
   const [modalOpen, setModalOpen] = useState(false);
-  const [_, dispatchScenarioUpdate] = useScenario();
+  const [, dispatchScenarioUpdate] = useScenario();
```
Comment (sprad) [inline src/page-multi-facility/ScenarioLibraryModal.tsx:117]: Apparently the linter isn't smart enough to consider variables names in a destructured array, but apparently you can also just not include the unused variable.

**Comment by sprad** [comment]
Pending final review, this should be mergeable now that the race condition has been fixed.  There are still launch blocking issues as defined in #168 , but this feature is behind a feature flag so those issues won't present themselves to users.

**Comment by macfarlandian** [comment]
nothing to add, LGTM!

**Pull request merged**